File: Swift/Leetcode.playground/Sources/24_DP/LC_740_DeleteAndEarn.swift
class Solution {
    func deleteAndEarn(_ nums: [Int]) -> Int {
        var map = [Int: Int]()
        var max_num = nums[0]
        for num in nums {
            map[num, default: 0] += 1
            max_num = max(max_num, num)
        }

        if max_num < 2 {
            return map[max_num, default: 0]
        }

        var dp = Array(repeating: 0, count: max_num + 1)
        dp[1] = map[1, default: 0] * 1
        for i in 2...max_num {
            dp[i] = max(dp[i - 1], dp[i - 2] + i * map[i, default: 0])
        }
        return dp[max_num]
    }
}

File: Java/24_DP/LC_740_DeleteAndEarn.java
class Solution {
    public int deleteAndEarn(int[] nums) {

        Map<Integer, Integer> map = new HashMap<>();
        int max = nums[0];
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
            max = Math.max(max, num);
        }

        int[] dp = new int[max + 1];
        dp[1] = map.getOrDefault(1, 0) * 1;

        for (int i = 2; i <= max; i++) {
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + i * map.getOrDefault(i, 0));
        }

        return dp[max];
    }
}

public class LC_740_DeleteAndEarn {

}
